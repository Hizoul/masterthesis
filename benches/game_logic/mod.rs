pub mod tetromino;
pub mod field;
pub mod transposition;