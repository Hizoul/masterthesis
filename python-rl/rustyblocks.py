#!/usr/bin/env python3
import math
import sys, ctypes
from os.path import abspath, dirname, realpath
from ctypes import c_char_p, c_uint32, c_uint8, c_int8, Structure, POINTER,c_void_p,c_int64,c_int16,c_double,c_uint64,c_char_p
import numpy as np
from numpy.ctypeslib import ndpointer
from PIL import Image


parent_folder = dirname(realpath(__file__)) + "/"
class Tuple(Structure):
    _fields_ = [("placed", c_uint8),("reward", c_double),("done", c_uint8),("winner", c_uint8)]
    def __str__(self):
        return "({},{},{},{})".format(self.placed, self.reward,self.done,self.winner)

class GameFieldS(Structure):
    pass

# libpath = "../target/debug/"
libpath = "../target/release/"
libname = "rustyblocks"
prefix = {'win32': ''}.get(sys.platform, 'lib')
extension = {'darwin': '.dylib', 'win32': '.dll'}.get(sys.platform, '.so')
absolute_lib_path = abspath(parent_folder + libpath + prefix + libname + extension)
print("Loading rustyblocks from ", absolute_lib_path)
rustLib = ctypes.cdll.LoadLibrary(absolute_lib_path)

rustLib.field_new.restype = POINTER(GameFieldS)
rustLib.field_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_new.restype = POINTER(GameFieldS)
rustLib.field_clone.argtypes = (POINTER(GameFieldS), )
rustLib.field_clone.restype = POINTER(GameFieldS)
rustLib.hex_free.argtypes = (POINTER(GameFieldS), )
rustLib.hex_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.hex_do_action_with_answer.restype = Tuple
rustLib.hex_to_array.restype = ndpointer(dtype=ctypes.c_uint8, shape=(121,))
rustLib.free_hex_array.argtypes = (ndpointer(dtype=ctypes.c_uint8, shape=(121,)), )
rustLib.hex_reset.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_string.restype = c_void_p
rustLib.field_to_json_log.argtypes = (POINTER(GameFieldS), )
rustLib.field_to_json_log.restype = c_void_p
rustLib.field_restore_log.argtypes = (POINTER(GameFieldS), c_char_p)
rustLib.field_to_array.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_to_array.restype = ndpointer(dtype=c_int8, shape=(420,))
rustLib.free_field_array.argtypes = (ndpointer(dtype=c_int8, shape=(420,)), )
rustLib.field_get_score.argtypes = (POINTER(GameFieldS),)
rustLib.field_get_score.restype = ndpointer(dtype=c_int16, shape=(2,))
rustLib.free_score_array.argtypes = (ndpointer(dtype=c_int16, shape=(2,)), )
rustLib.field_do_action.argtypes = (POINTER(GameFieldS), c_uint8)
rustLib.field_do_action.restype = c_uint8
rustLib.field_do_action_with_answer.argtypes = (POINTER(GameFieldS), c_uint8, c_uint8)
rustLib.field_do_action_with_answer.restype = Tuple
rustLib.field_do_action_with_answer_heu.argtypes = (POINTER(GameFieldS), c_double,c_double,c_double,c_double, c_uint8)
rustLib.field_do_action_with_answer_heu.restype = Tuple
rustLib.field_do_action_self_play.argtypes = (POINTER(GameFieldS), c_uint8, c_int8)
rustLib.field_do_action_self_play.restype = Tuple
rustLib.field_is_game_over.argtypes = (POINTER(GameFieldS), )
rustLib.field_is_game_over.restype = c_uint8
rustLib.field_get_reward.argtypes = (POINTER(GameFieldS), )
rustLib.field_get_reward.restype = c_double
rustLib.field_counter_action.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action.restype = c_int64
rustLib.field_counter_action_index.argtypes = (POINTER(GameFieldS), )
rustLib.field_counter_action_index.restype = c_uint8
rustLib.field_reset.argtypes = (POINTER(GameFieldS), )
rustLib.free_string.argtypes = (c_void_p, )
rustLib.eval_heuristic_weights.argtypes = (c_double,c_double,c_double,c_double)
rustLib.eval_heuristic_weights.restype = c_uint64

def field_to_string(field):
  ptr = rustLib.field_to_string(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)
def field_to_log(field):
  ptr = rustLib.field_to_json_log(field)
  try:
    return ctypes.cast(ptr, ctypes.c_char_p).value.decode('utf-8')
  finally:
    rustLib.free_string(ptr)

def field_to_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy().reshape((42,10))
  rustLib.free_field_array(original_array)
  return result

def field_score(field):
  original_array = rustLib.field_get_score(field)
  result = original_array.copy()
  rustLib.free_score_array(original_array)
  return result

def field_to_cnn_array(field, self_play_is_second_player):
  original_array = rustLib.field_to_array(field, self_play_is_second_player)
  result = original_array.copy()
  newres = np.zeros(shape=(32,32,3), dtype=np.uint8)
  for i in range(0, 200):
    y = math.floor(i / 10)
    x = i % 10
    current = 0
    val = 0
    if result[i] == 1 or result[i] == 2:
      current = 0
      if result[i] == 1:
        val = 128
      else:
        val = 255
    elif result[i] == 3 or result[i] == 4:
      current = 2
      if result[i] == 3:
        val = 128
      else:
        val = 255
    newres[y][x][current] = val
  for i in range(200, 210):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i] + 124
    newres[y][x][1] = val
  for i in range(210, 230):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    elif val > 0:
      val = val * 50
    newres[y][x][1] = val
  for i in range(230, 420):
    y = math.floor((i-200) / 10)
    x = (i % 10) + 10
    val = result[i]
    if val == -1:
      val = 0
    else:
      val = 255
    newres[y][x][1] = val
  rustLib.free_field_array(original_array)
  return newres

def hex_to_array(field):
  original_array = rustLib.hex_to_array(field)
  result = original_array.copy().reshape((11,11))
  rustLib.free_hex_array(original_array)
  return result

def field_string_to_array(field_string):
  field_array = np.zeros((20,10), dtype=np.uint8)
  row = 0
  col = 0
  for i in range(len(field_string)):
    entry = field_string[i]
    if entry == '\n':
      col = 0
      row += 1
    elif row < 20 and col < 10:
      field_array[row][col] = ord(entry) - 48
      col += 1
  return field_array

def save_img(data, name):
  Image.fromarray(data, 'RGB').save(name)