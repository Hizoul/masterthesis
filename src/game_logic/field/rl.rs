use super::GameField;
use crate::heuristics::{heuristic_value_enemy_block,heuristic_touching,heuristic_value_connectability};
use rayon::prelude::*;
use rand::Rng;
use std::sync::RwLock;
use std::f64::MIN;
use rand::{thread_rng};

pub trait LearnHelper {
  fn update_reward(&mut self, player: i8, include_difference: bool) -> f64;
  fn basic_reward(&mut self, player: i8) -> f64;
  fn calculate_full_reward(&mut self, player: i8, include_difference: bool) -> f64;
  fn get_random_heuristic_play(&mut self) -> u8;
  fn get_best_heuristic_play(&mut self, true_randomness: bool) -> u8;
  fn get_best_heuristic_play_for_params(&mut self, connectability_weight: f64, enemy_block_wheight: f64, touching_weight: f64, score_weight: f64, true_randomness: bool, heuristic_type: u8) -> u8;
  fn get_other_player(&self, current_player: i8) -> i8;
}

impl LearnHelper for GameField {
  fn get_other_player(&self, current_player: i8) -> i8 {
    let next = current_player + 1;
    if next >= self.players {
      0
    } else {
      next
    }
  }
  fn update_reward(&mut self, player: i8, include_difference: bool) -> f64 {
    let mut new_reward = self.scores[player as usize] as f64; //(self.scores[player as usize] - self.scores[other as usize]) as f64;
    if include_difference {
      let other = self.get_other_player(player);
      new_reward -= self.scores[other as usize] as f64;
    }
    new_reward += heuristic_touching(self, player) * 0.3;
    new_reward /= 100.;
    new_reward
  }
  fn basic_reward(&mut self, player: i8) -> f64 {
    self.scores[player as usize] as f64
  }
  fn calculate_full_reward(&mut self, player: i8, include_difference: bool) -> f64 {
    let mut reward = 0.0;
    let indices = self.played_indices.clone();
    let mut shadow_field = self.empty_clone();
    for index in indices {
      let is_players_turn = shadow_field.current_player == player;
      shadow_field.place_block_using_play(index);
      if is_players_turn {
        reward += shadow_field.update_reward(player, include_difference);
      }
    }
    reward
  }
  fn get_random_heuristic_play(&mut self) -> u8 {
    let mut rng = thread_rng();
    self.get_best_heuristic_play_for_params(
      rng.gen_range(-1.0, 10.0),
      rng.gen_range(-1.0, 10.0),
      rng.gen_range(-1.0, 5.0),
      rng.gen_range(0.1, 25.0), true, 0)
  }
  fn get_best_heuristic_play(&mut self, true_randomness: bool) -> u8 {
    self.get_best_heuristic_play_for_params(0.3, 0.5, 1.0, 20.0, true_randomness, 0)
  }
  fn get_best_heuristic_play_for_params(&mut self, connectability_weight: f64, enemy_block_wheight: f64, touching_weight: f64, score_weight: f64, true_randomness: bool, heuristic_type: u8) -> u8 {
    let current_max_rwlock = RwLock::new(MIN);
    let player = self.current_player;
    let to_copy_from = &self;
    let best_index_rwlock: RwLock<Vec<u8>> = RwLock::from(Vec::with_capacity(175));
    (0..self.possible_plays.len()).into_par_iter().for_each(|current_index| {
      let mut field = to_copy_from.empty_clone();
      field.copy_from(to_copy_from);
      field.place_block_using_play(current_index as u8);
      let heuristic_value: f64 = match heuristic_type {
        1 => heuristic_touching(&field, player) + field.scores[player as usize] as f64,
        _ => (heuristic_value_connectability(&field, player) * connectability_weight)
      + (heuristic_value_enemy_block(&field, player) * enemy_block_wheight) + (heuristic_touching(&field, player) * touching_weight) +
      (field.scores[player as usize] as f64 * score_weight)
      };
      let mut current_max_writeable = current_max_rwlock.write().unwrap();
      let currmaxcmp = *current_max_writeable;
      if heuristic_value > currmaxcmp {
        let mut best_index_writeable = best_index_rwlock.write().unwrap();
        best_index_writeable.clear();
        best_index_writeable.push(current_index as u8);
        *current_max_writeable = heuristic_value;
      } else if heuristic_value == currmaxcmp {
        let mut best_index_writeable = best_index_rwlock.write().unwrap();
        best_index_writeable.push(current_index as u8);
      }
    });
    let mut best_index_writeable = best_index_rwlock.write().unwrap();
    best_index_writeable.sort();
    if best_index_writeable.len() == 0 {
      0 
    } else if best_index_writeable.len() == 1 {
      best_index_writeable[0]
    } else {
      best_index_writeable[if true_randomness {rand::thread_rng().gen_range(0, best_index_writeable.len()-1)} else {self.rng.gen_range(0, best_index_writeable.len()-1)}]
    }
  }
}