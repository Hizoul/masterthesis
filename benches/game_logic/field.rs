extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::log::{GameLog};
use rustyblocks::game_logic::cache::{ShapeCache};
use std::sync::{Arc};
use rand::prelude::*;

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":0,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":0,\"x\":7,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":9,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":10,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":8,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":7,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":4,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":5,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":4,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":8,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":7,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":9,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":0,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":4,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":1,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":4}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":5,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":2,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":3,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":0,\"y\":10}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":7,\"y\":18}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";


struct Foo {
  counter: usize
}

impl Foo {
  pub fn add(&mut self) {self.counter+=1;}
  pub fn get(&self) -> usize {self.counter}
}

trait FooTrait {
  fn t_add(&mut self);
  fn t_get(&self) -> usize;
}

impl FooTrait for Foo {
  fn t_add(&mut self) {self.add();}
  fn t_get(&self) -> usize {self.get()}
}

pub fn bench_field(c: &mut Criterion) {
  let cache = ShapeCache::new();
  let rules = GameRules::default();
  let p = Arc::new(cache);
  let p2 = p.clone();
  let p3 = p.clone();
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  c.bench_function("deterministic_play_until_over", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p.clone(), rules);
    while !f.game_over {
      f.place_block_using_play(0);
    }
  }));
  c.bench_function("random_play_until_over", move |b| b.iter(|| {
    let mut  f = GameField::new_with_cache(2, p2.clone(), rules);
    while !f.game_over {
      let p = f.possible_plays.len() as u8;
      f.place_block_using_play(rand::thread_rng().gen_range(0,p));
    }
  }));
  c.bench_function("replay_log", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p3.clone(), rules);
    f.restore_from_log(&log, false);
  }));
  c.bench_function("traitcost_direct", move |b| b.iter(|| {
    let mut f = Foo {counter:0};
    for _ in 0..99999999 {
      while f.get() < 9999999999 {
        f.add();
      }
    }
  }));
  c.bench_function("traitcost_struct", move |b| b.iter(|| {
    let mut f: Box<dyn FooTrait> = Box::new(Foo {counter:0});
    for _ in 0..99999999 {
      while f.t_get() < 9999999999 {
        f.t_add();
      }
    }
  }));
}