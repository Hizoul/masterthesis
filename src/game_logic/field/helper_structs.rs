use serde::{Deserialize, Serialize};
use crate::game_logic::tetromino::{Tetromino};
use std::sync::{Arc};

#[derive(Debug,Clone)]
pub struct Touching {
  pub block: Arc<Tetromino>,
  pub touching: Vec<Arc<Tetromino>>,
  pub connected: Vec<Arc<Tetromino>>
}

impl PartialEq for Touching {
  fn eq(&self, other: &Touching) -> bool {
    self.block == other.block
  }
}

impl Touching {
  pub fn new(block: Arc<Tetromino>) -> Touching {
    return Touching {
      block, touching: Vec::with_capacity(4), connected: Vec::with_capacity(4)
    }
  }
}

#[derive(Debug,Serialize,Deserialize,Clone)]
pub struct OwnedPosition {
    pub x: i8,
    pub y: i8,
    pub from: i8
}

#[derive(Debug,Serialize,Deserialize,Copy,Clone)]
pub struct GameRules {
  pub roll_dice: bool,
  pub limit_minus: bool
}

impl GameRules {
  pub fn original() -> GameRules {
    GameRules {roll_dice:true,limit_minus:true}
  }
  pub fn original_without_limit() -> GameRules {
    GameRules {roll_dice:true,limit_minus:false}
  }
  pub fn deterministic() -> GameRules {
    GameRules {roll_dice:false,limit_minus:true}
  }
  pub fn deterministic_without_limit() -> GameRules {
    GameRules {roll_dice:false,limit_minus:false}
  }
}

impl Default for GameRules {
  fn default() -> GameRules {
    GameRules::original_without_limit()
  }
}