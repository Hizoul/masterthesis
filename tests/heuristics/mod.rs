use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::render::Renderable;
use rustyblocks::game_logic::log::{GameLog};
use rustyblocks::heuristics::{heuristic_value_width_territory,heuristic_value_height_territory,heuristic_value_all_territory,heuristic_value_enemy_block,heuristic_value_connectability};
use insta::{assert_debug_snapshot,assert_snapshot};
use rustyblocks::game_logic::field::rl::LearnHelper;
use rustyblocks::game_logic::field::helper_structs::{GameRules};

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":1,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":0,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":0,\"x\":5,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":2,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadConsidering\":{\"play_index\":0}}]}";

const CONNECTIVITY_TEST: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":3,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";

#[test]
fn heuristics_via_replay() {
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  let mut field = GameField::new(2);
  field.restore_from_log(&log, false);
  assert_snapshot!("field_after_restore", &field.to_field_string());
  assert_debug_snapshot!("lowest_ys", field.lowest_ys);
  assert_debug_snapshot!("width", (heuristic_value_width_territory(&field, 0),heuristic_value_width_territory(&field, 1)));
  assert_debug_snapshot!("height", (heuristic_value_height_territory(&field, 0),heuristic_value_height_territory(&field, 1)));
  assert_debug_snapshot!("full_territory", (heuristic_value_all_territory(&field, 0),heuristic_value_all_territory(&field, 1)));
  assert_debug_snapshot!("enemy_block", (heuristic_value_enemy_block(&field, 0),heuristic_value_enemy_block(&field, 1)));


  
  let log2: GameLog = serde_json::from_str(CONNECTIVITY_TEST).unwrap();
  let mut field2 = GameField::new(2);
  field2.restore_from_log(&log2, false);
  assert_snapshot!("field2_after_restore", &field.to_field_string());
  assert_debug_snapshot!("connectability", (heuristic_value_connectability(&field2, 0),heuristic_value_connectability(&field2, 1)));
  assert_debug_snapshot!("width2", (heuristic_value_width_territory(&field2, 0),heuristic_value_width_territory(&field2, 1)));
  assert_debug_snapshot!("height2", (heuristic_value_height_territory(&field2, 0),heuristic_value_height_territory(&field2, 1)));
  assert_debug_snapshot!("full_territory2", (heuristic_value_all_territory(&field2, 0),heuristic_value_all_territory(&field2, 1)));
  assert_debug_snapshot!("enemy_block2", (heuristic_value_enemy_block(&field2, 0),heuristic_value_enemy_block(&field2, 1)));
  assert_debug_snapshot!("rewards", (field2.update_reward(0, false),field2.update_reward(1, false)));
}

#[test]
fn heuristics_battle() {
  let mut field = GameField::new_with_rules(2, GameRules::deterministic());
  while !field.game_over {
    let play = field.get_best_heuristic_play(false);
    field.place_block_using_play(play);
  }
  assert_snapshot!("battle_end", &field.to_field_string());
  assert_debug_snapshot!("battle_scores", field.scores);
}
#[test]
fn random_heuristics_battle() {
  let mut field = GameField::new_with_rules(2, GameRules::deterministic());
  while !field.game_over {
    let play = field.get_random_heuristic_play();
    field.place_block_using_play(play);
  }
  assert_snapshot!("battle_end", &field.to_field_string());
  assert_debug_snapshot!("battle_scores", field.scores);
}