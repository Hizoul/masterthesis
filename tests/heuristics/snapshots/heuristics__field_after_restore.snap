---
created: "2019-09-30T08:33:20.110862641Z"
creator: insta@0.8.1
source: tests/heuristics/mod.rs
expression: "&field.to_field_string()"
---

0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0020000000
0020000000
0020000000
0010000000
0440000000
3410000000
0422222000
4342134440

